flow/README.md

```diff
@@ -341,18 +355,52 @@ you are holding the corresponding future.
     [T]
     ```
 
-TODO finish documenting/implementing the following.
-1. Vtables collected from default-constructed instances
-1. Requirements (serialize must be cheap for a default-constructed instance, must have a serialize method or implement a trait.)
-1. Traits/Concepts: vector_like, union_like, dynamic_size, scalar
-1. isDeserializing idiom
-1. Gotchas (serialize gets called more than once on save path, maybe more)
+1. Flatbuffers Traits
+
+    In order to serialize a type as a flatbuffers vector, struct, or union, you can implement the appropriate trait for your type.
+    - `scalar_traits` corresponds to a flatbuffers struct. See `UID` for an example.
+    - `vector_like_traits` corresponds to a flatbuffers vector. See `VectorRef` for an example.
+    - `dynamic_size_traits` corresponds to a flatbuffers vector of uint8_t. See `StringRef` for an example.
+    - `union_like_traits` corresponds to a flatbuffers union. See `boost::variant` for an example.
+
+1. Potential Gotchas
+    - Flatbuffers 'vtables' are collected from default-constructed instances of
+      each type. Consequently types serialized by flatbuffers should have cheap
+      default constructors. Future work: we may be able to collect vtables
+      without an instance of a type using `declval`.
+
+    - `T::serialize` may get called multiple times when serializing `T`. It is
+      guaranteed to be called only once for deserialization though, and thus
+      the `Ar::isDeserializing` idiom is appropriate. Future work: in theory we
+      don't need to call `T::serialize` multiple times when serializing, but
+      this would complicate the implementation.
+
 1. File identifiers
+
+    [File identifiers](https://google.github.io/flatbuffers/md__schemas.html)
+    are used to sanity check that the message you're deserializing is of the
+    schema you expect. You can give a type `T` a file identifier by making
+    `T::file_identifier` a static member of type `FileIdentifier`. You don't
+    need to change the file identifier for a type when evolving its schema.
```

Comment:
Maybe also mention how to define a file identifier for types we don't control?

Reply:
Done